### ISSUE TITLE
Add external elements to the city

### ISSUE THREAD

**Issue by jmespadero** [jmespadero/cityGen#7]
Place objects in the exterior of the city. For example Farms, roads, rivers, etc...

The ideas are:
1.  Build a random polyline that begins in the gate of the city and goes far away. Paste a texture of road in the polyline (similar to current streets).
2. Reuse the same algorithm in 1 to generate a random river near of the city.
3. Create an horizon line with a landscape with mountains (a [trompe l'oeil](https://en.wikipedia.org/wiki/Trompe-l%27%C5%93il) o trampantojo)
4. Generate a forest with some trees, stones, dry leaves, etc...
5. Some random building near of the city. (windmills, serreria, vigilance towers, etc...)

**Comment by jmespadero** [comment]
Hi @sergiofv93,

This is a tutorial explaining how to manually model a river in BGE
https://www.youtube.com/watch?v=AwLtwbmU74U

The idea is to extract the part in which he creates the water material and generate your own water material to use on Citygen. It is not necessary to animate the UV coordinates for now, because rivers are going to be outside of the cities...

Also, there is a Seamless "Dirt" texture that may be used to create a material for roads outside of the city.

**Comment by sergiofv93** [comment]
Hi @jmespadero,

I'm writting an algorithm for creating rivers and paths outside of the city.
Check the branch https://github.com/sergiofv93/cityGen/tree/riverAndTrail

Greetings.

**Comment by jmespadero** [comment]
Hi @sergiofv93 ,

Regard to how to build a planar mesh from a NURBS...

You cannot use a blender curve in BGE, because it only accepts objects of type mesh.
So, supposing that you have defined a river/road using a NURBS, you can do this steps.

![rivermesh svg](https://user-images.githubusercontent.com/6451497/37284149-c55c0812-25fa-11e8-8a45-b6421a9f8d28.png)

Repeat for each vertex in the polyline:
1) Evaluate your curve to create a 2D polyline in XY.
2) For each vertex, build an `tangent vector` that goes from previous vertex to next vertex.
3) Build a `normal vector` by rotating previous vector 90º. In 2D you can do this easily by swapping the X and Y coordinates of the vector and change sign of Y coordinate
4) Change length of normal vector by normalization+ multiply by "width of the river/2". Compute a vertex of the mesh as `current coordinate` + `normal vector`
5) Compute opposite vertex in the mesh as `current coordinate` - `normal vector`

You will get twice the vertex than the original 2D polyline. Just connect them in the right order to create a mesh.

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

I've pushed new content in the riverAndTrail branch.
I've deleted the first river algorithm, because as I've search on internet, if I want to extrude the river from a plane mesh the coordinates for each region to extrude are local, not global. So If I move a region to coordinates (3, 4, 0), the region doesn't locate in those coordinates. It moves 3 points on X axis and 4 points on Y axis.

The mesh function I'm using is:
`bpy.ops.mesh.extrude_region_move(TRANSFORM_OT_translate={"value": (3, 3, 0)})`

As we can find, it has several parameters, but the most important for me is TRANSFORM_OT_translate with the movement coordinates. I've seen the API and I don't find any parameter to asign values like "GLOBAL", "LOCAL" or something like that to convert those coordinates in global coordinates.

The problem is that I can't define points to extrude the mesh. I only can "show it the course it has to follow".

Any idea?

Thanks in advance,
greetings.

**Comment by jmespadero** [comment]
It not clear to me what you are trying to do. Are you trying to mimic the manual procedure of extruding a plane?

Any case, I do not recommend the use of bpy.ops.* operators if you can avoid then, due to high performance penalties in the generator. 

So my idea is still:

1. Generate a list of points with the "river skeleton". Draw with a polyline just fot debugging purposes.
2. Generate a list of points with the "river boundary". Probably you want to create two separate lists
 for left and right boundary of the river.
3. Create a mesh "knitting" the points of the boundary.

You have a working example in the method makeGround.
https://github.com/jmespadero/cityGen/blob/682f77b929f78aac99079c9446d8cd13681a3474/cityGen3D.py#L295

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

I've keep your instructions writting a new algorithm for create the river. Following the "skeleton" of the river, I get two symmetric lists of points for each side of the river. As you can see, I've drawn a cube in each point for debbugging. With this basic shape, now I will create the river mesh.

Thanks in advance,
greetings.

**Comment by sergiofv93** [comment]
**River Creation Algorithm**

For create the river, I debated with @jmespadero for select the best option, both in efficiency and complexity. We agree that I should use a 1D variant of the **Random Midpoint Displacement Fractal**, also know as [Diamond-square Algorithm](https://en.wikipedia.org/wiki/Diamond-square_algorithm).

So now I will write how I build the river step by step in several posts while I'm progressing. I will explain the details of this algorithm in the project documentation, not here.

**Comment by sergiofv93** [comment]
**Step 1: Create init and end points (A and B)**

I have the city radius, which I will use to calculate a distance from the coordinates origin to the river. So, for example, I think that choosing _cityRadius * 2_ as the "distance" is a good option.

The river is "distance" from the coordinates origin, and the length of the river will be _distance * 2_, it means that points A and B are separated _distance * 2_.

![step1](https://user-images.githubusercontent.com/22729865/37683327-90eccf06-2c8c-11e8-8dfd-3329cc02f456.png)

And for now, the A and B points will be the same for all executions. Maybe later I decide to change them and make it "random".

Finally, If I add a cube at point A and B for debug, we can see this blender scene:

![step1_2](https://user-images.githubusercontent.com/22729865/37683688-b34f4b36-2c8d-11e8-8900-afb0071b24d5.png)

**Comment by sergiofv93** [comment]
**Step 2: Calculate new river point**

The way to create a simple point of the river which is not the init or end point is to divide the distance between A and B and get the perpendicular vector. I won't explain the code details here, but in the documentation of the proyect.

When I've that vector, I'll calculate the location of the new point. In summary, I will get the middle point between A and B, and using the vector and a random distance, I'll locate the new point in a valid position.

![step2_1](https://user-images.githubusercontent.com/22729865/37741327-3aaedd3e-2d61-11e8-8df9-e71e53f72af4.png)


For this I use a parameter of the function, called "factor". This "factor" determines the perpendicular distance limit for locating the new point, and it's a percentage that will be multiplied by a random value between -1 and 1. It has to be between -1 and 1 because we want to allow the new river point can be located in both sides of the line that crosses A and B.

If I add cubes for the middle point and the new river point, we get this blender scene:

![step2_2](https://user-images.githubusercontent.com/22729865/37741346-5614e280-2d61-11e8-834a-d52278300faf.png)

In the next step, I will create a recursive function depending on the "resolution" the user (me) specifies.

**Comment by sergiofv93** [comment]
**Step 3: Create the recursive algorithm**

Finally, I've created a recursive algorithm to create all the points recursively. In general terms, I've needed to create a variable which controls the number of recursive levels the algorithm will execute. I've called this variable "resolution". If its value is 1, we'll have only 1 recursive level, if is 3, we'll have 3 recursive levels, etc.

![step3_1](https://user-images.githubusercontent.com/22729865/38217538-0dc32f1a-36cf-11e8-8699-dfbe4d5c0911.png)


Each level divides the distance between two points (A and B) in the middle and calls the next recursive level from the point A to the medium point and from the medium point to B.

The result, as we can see in the following screenshots, is that the river will have more vertices the higher the value of the variable "resolution" (resolution 1 and 3 in the screen captures):

![step3_2](https://user-images.githubusercontent.com/22729865/38218527-60889dc2-36d2-11e8-80d0-0441e374ed6f.png)  ![step3_4](https://user-images.githubusercontent.com/22729865/38218530-60c43b2a-36d2-11e8-812e-300fd3eb8d8f.png)

I've to comment that the smalest cubes are the medium point on each recursive level. And now I have my river skeleton. Now I'll have to calculate the mesh points.

**Comment by jmespadero** [comment]
Two small recommendations here:
1) You should use a tool to draw your diagrams in svg or png file format. It is worth to learn basic usage of [inkscape](https://inkscape.org) or libreoffice-draw or [draw.io](https://www.draw.io/) to create diagrams.
2) You should start to count resolution steps in 0, not 1. So the straight line that goes from A to B is the fractal of resolution "0", meaning you have applied the midpoint displacement recursion zero times. Also, the number of segments in your skeleton will be 2^n, which is a nice formula. You could also compute things as the "expected segment length" and "expected river length" with formulas that are simpler if you start counting in 0.

**Comment by sergiofv93** [comment]
Hi @jmespadero 

I've fix that issue and I start the resolution steps count in 0. Now, I've pushed some small changes:
I have ordered the list of points, before that, the points was been added in the "recursive" order, resulting a list like [A,E,D,F,C,H,G,I,B] and now I have the list [A,B,C,D,E,F,G,H,I]. It will be useful for the next "mesh points" creation. I fixed it appending the new points between de firts recursive step and the last recursive step in each recursive level:

```
newRiverPoint(p1, p3, factor, list, resolution - 1)
list.append(p3)
newRiverPoint(p3, p2, factor, list, resolution - 1)
```

Thanks for your advice,
greetings.

**Comment by jmespadero** [comment]
A small hack to temporally plot the skeleton...

Use the [duplicateAlongSegment](https://github.com/sergiofv93/cityGen/blob/3d4a9a776372008e597af0c69d0b9f20821cd175/cityGen3D.py#L84) method to plot the skeleton segment-by-segment.

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

I have keep your advice and now I can draw the skeleton with a sand trail. The next step will be to draw the both sides of the river. Here we have a screen capture:

![step3_5](https://user-images.githubusercontent.com/22729865/38506710-ebe10a1a-3c1a-11e8-9bf9-428acd95e6d0.png)

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

Finally, I have created the both sides of the river and I have drawn them, as you can see in the picture. Next step is to create the mesh, but I'll do that tomorrow. Now I will try to install and learn how to use **Inkscape**, as you said me. Here we have the screen capture:

![step3_6](https://user-images.githubusercontent.com/22729865/38509104-45d92916-3c21-11e8-9ea1-d558f5671cd6.png)

**Comment by sergiofv93** [comment]
Hi @jmespadero ,

Finally, I've completed the **River Creation Algorithm**. Using the from_pydata function of your code, I've created the mesh given 2 lists of points. With this 2 lists of points, I've created a new list containing all those points in order. Then I've implemented a new list of faces which will be useful to call the from_pydata function.
When I had those steps, I've created a new material called "Water" to add to the river mesh, and finally I've normalized the river width, which I can control passing a parameter called "width" to the function.

Now, I have the next results:

![step4_1](https://user-images.githubusercontent.com/22729865/38686184-0d84d948-3e74-11e8-9968-8a1739bcb48e.png)

Now, I need to finish the customization of the material. It has to act like a mirror of some montains and present a "real water texture". Then, I'll begin to create the trail from the principal door of the city.

Kind Regards.

**Comment by jmespadero** [comment]
I would suggest to take a screen-shot after selecting the river, and then enter in "edit" mode to show the underlying triangle structure of the river mesh.
![rivertriangles](https://user-images.githubusercontent.com/6451497/38689316-9415be12-3e7b-11e8-8216-4c2f1f4dad08.jpg)

**Comment by sergiofv93** [comment]
Here we have it:

![captura de pantalla de 2018-04-12 18-08-44](https://user-images.githubusercontent.com/22729865/38689701-9284b520-3e7c-11e8-8e8c-02d85b8b6e59.png)

**Comment by sergiofv93** [comment]
What's more, has been quite easy to create the trail using the functions for rivers. Only have to calculate the origin and end points (gate and end) and give it a different texture.

![captura de pantalla de 2018-04-12 18-18-24](https://user-images.githubusercontent.com/22729865/38690318-283b78c8-3e7e-11e8-8ed5-f30592909a11.png)


What do you think about it, @jmespadero ?

**Comment by jmespadero** [comment]
I think this task is almost finished :tada: 

Now, let's refine the water material for the river...

As said at the beginning of this thread, there is a nice tutorial explaining how to manually [model a river in BGE](https://www.youtube.com/watch?v=AwLtwbmU74U). Currently you have the demo in your repository in the file water.blend.

You can just reuse the water material in water.blend and put it into cg-library.blend, all textures are in CC-zero license. You need to adapt the Material->Texture->Mapping->Coordinates to be "Global" instead "UV" (he used UV to nicely animate the water, but it is out of our scope)
You probably will have to change the "Mapping->Size" parameters to adapt the size of the ripples to our own scale.

**Comment by sergiofv93** [comment]
**Closure specifications modified.**

In contrast to what we have talked first time, this Issue can be closed without adding a mountains line outside the city or adding windmills or external buildings. The work done in the river and trail creations are enough to close this issue.


**River**

Finally, I've completed the river creation and customization by upgrading the texture of the water, for giving a more realistic look. Details about it will be written in the documentation. Here we have a screen capture of the results:

![step5_1](https://user-images.githubusercontent.com/22729865/38821834-efccb688-41a1-11e8-831f-9ddb2bbdede4.png)


**Trail**

The trail has been created using the same functionality of the river, but the texture applied changes. This time, I've used a texture sand like, that I've found on the Internet. Another time, the details about it will be written in the documentation. You can see the results in the next screen capture:

![step5_2](https://user-images.githubusercontent.com/22729865/38822544-0dac4d74-41a4-11e8-91ce-98fbb6396d4d.png)

**Additional changes**

I've improved the performance of the code. Several functions were redundant, now they are unified in only 3 functions that control the creation of the polyline and the adding of the meshes and textures. Details about it will be written in the documentation of the proyect.

Thanks for your help @jmespadero

**Comment by jmespadero** [comment]
I have merged your #17 Pull-Request, I think that it is enough work for the outside of the city.

Congratulation for this result.

Nice ideas for (probably other guy) extend external elements...
* Add a lagoon -> Large patch of terrain with you water material
* Add the sea -> Large patch of terrain near of the skydome with a modified water material.
* Generate a forest with some trees, stones, dry leaves, etc... 
* Some random building near of the city. (windmills, serreria, vigilance towers, etc...)
